### PR TITLE
Process.WaitForExit: don't wait for standard streams when process was killed.

### PR DESCRIPTION
**Pull request by tmds**
This is a common pattern:

```
process.Kill();
process.WaitForExit();
```

When the process has redirected standard output or standard error
`WaitForExit` will block until all descendants of the child process (that
inherited these streams) have also terminated.

This change will make `WaitForExit` no longer wait for these descendants
when the process was killed first.

Additionally, when the user has cancelled the reading by calling
`CancelOutputRead` and `CancelErrorRead`, `WaitForExit` will no longer
wait for the streams either.

@danmoseley @stephentoub @adamsitnik @eiriktsarpalis ptal

### CODE REVIEW

**Comment by stephentoub** [comment]
What's the reason for he proposed change?

**Comment by tmds** [comment]
> What's the reason for the proposed change?

To better match with expectation that `WaitForExit` will return timely after a successful `Kill`.

The current behavior can introduce unexpected long blocking.

Users that are aware call `WaitForExit` with a timeout. If they pick the timeout that is 'too small', it may return before the process terminates and cause some unexpected behavior that occurs very infrequently and is hard to trace back.

**Comment by tmds** [comment]
The tests aren't passing on Windows.

> Assert.Equal() Failure\r\n          ↓ (pos 0)\r\nExpected: Sleep child started\r\nActual:   The system cannot find the file specified···\r\n          ↑ (pos 0)

I think it doesn't find "sleep".

**Comment by danmoseley** [comment]
Is the current behavior the same on both Windows and Unix? What is the .NET Framework behavior? If we are matching that, it seems this change could be significantly breaking.

**Comment by tmds** [comment]
Behavior is the same on Windows and Unix.

The change is meant to be non-breaking:
The `WaitUntilEOF` in `WaitForExit` is meant as a way to ensure we've consumed all stdout/stderr from the app. We're using the `Kill` call from the user as an indication he's no longer interested.
For backwards-compatibility, we're faking a final `null` `DataReceived` event.

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs [hunk 1]
```diff
@@ -67,7 +74,17 @@ internal AsyncStreamReader(Stream stream, Action<string?> callback, Encoding enc
...
+            if ((state != OperationStateNone) &&
+                (state != OperationStateCanceled))
+            {
+                return;
+            }
+
+            if (Interlocked.CompareExchange(ref _operationState, OperationStateReading, state) != state)
+            {
+                return;
+            }
```
Comment (tmds) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs:87]: Why does `BeginReadLine` ever makes calls to `FlushMessageQueue`?
I think that is `ReadBufferAsync`'s responsibility?

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 2]
```diff
@@ -2378,6 +2378,145 @@ public void Start_PassesArgumentsList_WhichGetsEscaped()
...
+            {
+                childOutputTcs.TrySetResult(e.Data);
+                if (e.Data == null)
+                {
+                    eofTcs.SetResult(null);
+                }
+            };
+            process.BeginOutputReadLine();
+            process.BeginErrorReadLine();
+            string childOutput = await childOutputTcs.Task.TimeoutAfter(30_000);
```
Comment (carlossanlop) [inline src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs:2422]: @tmds the `TimeoutAfter` extension method has been removed in [this PR](https://github.com/dotnet/runtime/pull/48842/files).

You will have to convert it to a `WaitAsync` extension method invocation instead.

File: src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 3]
```diff
@@ -2378,6 +2378,145 @@ public void Start_PassesArgumentsList_WhichGetsEscaped()
...
+            }
+            else
+            {
+                process.CancelOutputRead();
+                process.CancelErrorRead();
+            }
+
+            Task asyncWaitTask = useAsyncAPI ? process.WaitForExitAsync() :
+                                               Task.Run(() => process.WaitForExit());
+            await asyncWaitTask.TimeoutAfter(30_000);
```
Comment (carlossanlop) [inline src/libraries/System.Diagnostics.Process/tests/ProcessTests.cs:2437]: The `sleep` child is meant to quit after 10 minutes. Is there a point in waiting 30 seconds? Seems like a lot.

**Review by carlossanlop — COMMENTED** [review]
Thanks for your PR, @tmds. I left some feedback for you to consider.

@Jozkee @adamsitnik let's talk about this PR in our next triage meeting to discuss any potential unintended breaking changes, and if we are ok to take them.

**Comment by jeffhandley** [comment]
@carlossanlop, @Jozkee, @adamsitnik -- please take a look at this in your next triage meeting to decide if we should take the behavioral change.

**Comment by jeffhandley** [comment]
@adamsitnik, @carlossanlop, @Jozkee -- merge conflicts have emerged on this, but we should decide if we want to take the behavioral change before we both resolving them.

**Comment by jeffhandley** [comment]
@adamsitnik This PR is assigned to you for follow-up/decision before the RC1 snap.

**Comment by stephentoub** [comment]
@jeffhandley, @adamsitnik, what is the plan here?  Last comment is from July talking about doing something with this prior to the RC1 snap, which has already happened.

**Review by adamsitnik — COMMENTED** [review]
Hello @tmds 

First of all, please accept my apologies for a very big delay.

To be honest with you, I am always hesistant to review PRs that try to change something non-trivial without disucssing it first. 
If we discuss things first, and agree that given scenario needs to be fixed, you won't waste your time by working on a fix. As a maintainer I am going to have some context (the review will be easier) and I am also going to have the chance to see what other users think about it (the more upvotes, the higher priority on my TODO list).
If we don't and it turns out that current implementation works as expected, we are both not happy about the outcome: you have spent time on a change that won't get merged, while I have to reject a PR knowing that someone had good intentions and spent some time working on it.

In this particular PR we have two changes:

> make WaitForExit no longer wait for these descendants
when the process was killed first

Currently, we wait for these descendants only when the user does not provide any timeout value, meaning that the user can wait up to infinity. I don't believe that this behavior should be changed. It's by design.

https://github.com/dotnet/runtime/blob/ef857626c5a0ad29cffb61a2f8db87e083ca2332/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs#L183-L188

https://github.com/dotnet/runtime/blob/ef857626c5a0ad29cffb61a2f8db87e083ca2332/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs#L213-L217

> when the user has cancelled the reading by calling
CancelOutputRead and CancelErrorRead

I am suprised that Cancel methods don't actualy cancel anything. It seems that we set the flag to true:

https://github.com/dotnet/runtime/blob/ef857626c5a0ad29cffb61a2f8db87e083ca2332/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs#L83-L85

But don't stop reading:

https://github.com/dotnet/runtime/blob/ef857626c5a0ad29cffb61a2f8db87e083ca2332/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs#L218-L219

The actual async read operation cancellation is requested by `Dispose` method:

https://github.com/dotnet/runtime/blob/ef857626c5a0ad29cffb61a2f8db87e083ca2332/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs#L256-L258

Would it not be simpler to just actually use the `_cancelOperation` flag in all `while(true)` loops and call `_cts.Cancel();` from `CancelOperation()`?

But even with that, I am not sure if the stream used by async reader supports cancellation as on Windows we pass `isAsync: false` to the `FileStream` ctor:

https://github.com/dotnet/runtime/blob/ef857626c5a0ad29cffb61a2f8db87e083ca2332/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs#L639

May I ask you to: create a new issue that describes the cancellation problem and send a new PR that addresses that? Once we have that PR, I am going to take a look at the Windows part and ensure that we are using a stream that supports cancellation. Once we get there, the issue can be closed (and the fix included in .NET 7).

**Comment by tmds** [comment]
> Currently, we wait for these descendants only when the user does not provide any timeout value, meaning that the user can wait up to infinity. I don't believe that this behavior should be changed. It's by design.

By design means this is intentional.
If a user kills a process, why would he then want to wait up to infinity to read its output?

This is how I came to making this PR:

msbuild has some code like this:

```cs
process.Kill();
bool exited = process.WaitForExit(timeout);
```

The timeout is here to avoid the the infinite wait when a grand child still holds the terminal.

The msbuild repo also had a long standing open issue where things went wrong for an unclear reason that was hard to reproduce.

The root cause was a bug in the `exited` `false` path, which was very unlikely, but not impossible, to occur.

**Pull request closed without merging**